### PR TITLE
[SPARK-19791] [ML] Add doc and example for fpgrowth

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?

Add a new section for fpm
Add Example for FPGrowth in scala and Java

updated: Rewrite transform to be more compact.

## How was this patch tested?

local doc generation.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala [hunk 1]
```diff
@@ -240,12 +240,13 @@ class FPGrowthModel private[ml] (
     val predictUDF = udf((items: Seq[_]) => {
       if (items != null) {
         val itemset = items.toSet
-        brRules.value.flatMap(rule =>
-          if (items != null && rule._1.forall(item => itemset.contains(item))) {
+        brRules.value.flatMap { rule =>
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: Nit, while we're here -- why change this bit?
Or if simplifying, what about

```
brRules.value.filter(_._1_forall(itemset.contains)).flatMap(_._2.filter(!itemset.contains(_)))
```

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * An example demonstrating FP-Growth.
+ * Run with
+ * {{{
+ * bin/run-example ml.FPGrowthExample
+ * }}}
+ */
+object FPGrowthExample {
+
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: nit: remove blank line

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  def main(args: Array[String]): Unit = {
+
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: nit: remove blank line

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    val spark = SparkSession
+      .builder
+      .appName(s"${this.getClass.getSimpleName}")
+      .getOrCreate()
+    import spark.implicits._
+
+    // $example on$
+    // Loads data.
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: I think this comment is pretty obvious, you can probably remove it

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    val dataset = spark.createDataset(Seq(
+      "1 2 5",
+      "1 2 3 5",
+      "1 2")
+    ).map(t => t.split(" ")).toDF("features")
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: I think it is better to explicitly declare the data instead of manipulating strings, that way it is very clear what the input data is for the example.  like this
```
val dataset = spark.createDataset(Seq(
      Array(1, 2, 5),
      Array(1, 2, 3, 5),
      Array(1, 2))
    ).toDF("features")
```

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    // Trains a FPGrowth model.
+    val fpgrowth = new FPGrowth().setMinSupport(0.5).setMinConfidence(0.6)
+    val model = fpgrowth.fit(dataset)
+
+    // get frequent itemsets.
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: should say "Display frequent itemsets."

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    model.freqItemsets.show()
+
+    // get generated association rules.
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: same as comment above

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    model.associationRules.show()
+
+    // transform examines the input items against all the association rules and summarize the
+    // consequents as prediction
+    model.transform(dataset).show()
+
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: nit: remove blank line

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala [hunk 9]
```diff
@@ -56,8 +56,8 @@ private[fpm] trait FPGrowthParams extends Params with HasFeaturesCol with HasPre
   def getMinSupport: Double = $(minSupport)
 
   /**
-   * Number of partitions (>=1) used by parallel FP-growth. By default the param is not set, and
-   * partition number of the input dataset is used.
+   * Number of partitions (positive) used by parallel FP-growth. By default the param is not set,
```
Comment (BryanCutler) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: why change to "positive", I think it was clearer before

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    // $example on$
+    // Loads data.
+    val dataset = spark.createDataset(Seq(
+      "1 2 5",
+      "1 2 3 5",
+      "1 2")
+    ).map(t => t.split(" ")).toDF("features")
+
+    // Trains a FPGrowth model.
```
Comment (BryanCutler) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: nit: technically it is just the line below calling `fit` that trains the model, I would move this comment down or just take it out

**Review by BryanCutler — COMMENTED** [review]
Just some minor comments.  I also ran the examples and output looks good.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 9 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: I presume to fix a javadoc error because angle brackets are read as opening an HTML tag

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 1 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: The change was about a style comment from the orginal PR that I missed. But it's great to see your suggestion. I'll run some test to confirm the performance.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 9 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: Yes. That's the reason. But still I'm getting some java doc error after merging code. Looking into it..

File: examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala (hunk 3 above)
Comment (hhbyyh) [inline examples/src/main/scala/org/apache/spark/examples/ml/FPGrowthExample.scala]: Thanks, I'll remove this line.

**Comment by hhbyyh** [comment]
ping @jkbradley since we're changing the FPGrowth `transform`.
Sean made a great suggestion to simplify `transform` code.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 9 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: I don't think the error is related to this change.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: Let's just use ">=1" but figure out how to escape the characters for javadoc.  We'll want to do that long-term.

**Comment by jkbradley** [comment]
The updated transform looks good; thanks for pinging!

**Comment by hhbyyh** [comment]
Refined some comments and minor things. This should be ready for review. Thanks.

**Comment by hhbyyh** [comment]
Please hold on merging this until https://github.com/apache/spark/pull/17321 is resolved.

**Comment by jkbradley** [comment]
I'll be happy to help get this merged now that the column renaming is done

**Comment by jkbradley** [comment]
Noting here: Please check out the "Issue this PR brought up" here: https://github.com/apache/spark/pull/17218

It may affect this PR.  Thanks!

**Comment by hhbyyh** [comment]
Updated with Python example.

File: docs/ml-frequent-pattern-mining.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+* `numPartitions`: the number of partitions used to distribute the work. By default the param is not set, and
+  partition number of the input dataset is used.
+
+The `FPGrowthModel` provides:
+
+* `freqItemsets`: frequent itemsets in the format of DataFrame("items"[Array], "freq"[Long])
+* `associationRules`: association rules generated with confidence above `minConfidence`, in the format of 
+  DataFrame("antecedent"[Array], "consequent"[Array], "confidence"[Double]).
+* `transform`: The transform method examines the input items in `itemsCol` against all the association rules and
+  summarize the consequents as prediction. The prediction column has the same data type as the
```
Comment (BenFradet) [inline docs/ml-frequent-pattern-mining.md]: I don't think this really explains what `transform` does or maybe it's just me?

I would have said something like:

>The transform method will produce `predictionCol` containing all the consequents of the association rules containing the items in `itemsCol` as their antecedents. The prediction column...

File: docs/ml-frequent-pattern-mining.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+PFP distributes the work of growing FP-trees based on the suffices of transactions,
+and hence more scalable than a single-machine implementation.
+We refer users to the papers for more details.
+
+`spark.ml`'s FP-growth implementation takes the following (hyper-)parameters:
+
+* `minSupport`: the minimum support for an itemset to be identified as frequent.
+  For example, if an item appears 3 out of 5 transactions, it has a support of 3/5=0.6.
+* `minConfidence`: minimum confidence for generating Association Rule. The parameter will not affect the mining
+  for frequent itemsets,, but specify the minimum confidence for generating association rules from frequent itemsets.
```
Comment (BenFradet) [inline docs/ml-frequent-pattern-mining.md]: also, there are two commas after itemsets

**Review by BenFradet — COMMENTED** [review]
a few remarks regarding the documentation otherwise it looks good :+1:

File: docs/ml-frequent-pattern-mining.md (hunk 11 above)
Comment (hhbyyh) [inline docs/ml-frequent-pattern-mining.md]: Thanks for the suggestion. I do wish to have a better illustration here. But the two containing in your version make it not that straightforward, and actually it should be items in `itemsCol ` contains the antecedents for association rules. 

I extend it to a longer version,

For each record in `itemsCol`,  the `transform` method will compare its items against the antecedents of each association rule. If the record contains all the antecedents of a specific association rule, the rule will be considered as applicable and its consequents will be added to the prediction result.  The `transform` method will summarize the consequents from all the applicable rules as prediction.

Comment (BenFradet) [inline docs/ml-frequent-pattern-mining.md]: even better :+1:

File: examples/src/main/python/ml/fpgrowth_example.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    # $example on$
+    df = spark.createDataFrame([
+        (0, [1, 2, 5]),
+        (1, [1, 2, 3, 5]),
+        (2, [1, 2])
+    ], ["id", "items"])
+
+    fpGrowth = FPGrowth(itemsCol="items", minSupport=0.5, minConfidence=0.6)
+    fpGrowthModel = fpGrowth.fit(df)
+
```
Comment (zero323) [inline examples/src/main/python/ml/fpgrowth_example.py]: Can we  `associationRules` example? After all this is the biggest advantage for the Python user.

Comment (hhbyyh) [inline examples/src/main/python/ml/fpgrowth_example.py]: definitely. Thanks.

**Comment by felixcheung** [comment]
right, how are we on this? let's get this ready soon and merge?
could you also add reference to the R example which is merged yesterday.

File: docs/ml-frequent-pattern-mining.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+Refer to the [Java API docs](api/java/org/apache/spark/ml/fpm/FPGrowth.html) for more details.
+
+{% include_example java/org/apache/spark/examples/ml/JavaFPGrowthExample.java %}
+</div>
+
+<div data-lang="python" markdown="1">
+Refer to the [Python API docs](api/python/pyspark.ml.html#pyspark.ml.fpm.FPGrowth) for more details.
+
+{% include_example python/ml/fpgrowth_example.py %}
+</div>
```
Comment (felixcheung) [inline docs/ml-frequent-pattern-mining.md]: add R please

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: docs/ml-frequent-pattern-mining.md (hunk 14 above)
Comment (hhbyyh) [inline docs/ml-frequent-pattern-mining.md]: Sure. Added reference to R example. Manually checked on generated doc.

File: docs/ml-frequent-pattern-mining.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  from frequent itemsets.
+* `numPartitions`: the number of partitions used to distribute the work. By default the param is not set, and
+  number of partitions of the input dataset is used.
+
+The `FPGrowthModel` provides:
+
+* `freqItemsets`: frequent itemsets in the format of DataFrame("items"[Array], "freq"[Long])
+* `associationRules`: association rules generated with confidence above `minConfidence`, in the format of 
+  DataFrame("antecedent"[Array], "consequent"[Array], "confidence"[Double]).
+* `transform`: For each transaction in itemsCol, the `transform` method will compare its items against the antecedents
```
Comment (felixcheung) [inline docs/ml-frequent-pattern-mining.md]: `itemsCol`?

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala [hunk 16]
```diff
@@ -82,8 +81,8 @@ private[fpm] trait FPGrowthParams extends Params with HasPredictionCol {
   def getNumPartitions: Int = $(numPartitions)
 
   /**
-   * Minimal confidence for generating Association Rule.
-   * Note that minConfidence has no effect during fitting.
+   * Minimal confidence for generating Association Rule. MinConfidence will not affect the mining
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: lower case `minConfidence`?

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala [hunk 17]
```diff
@@ -268,12 +269,8 @@ class FPGrowthModel private[ml] (
...
       if (items != null) {
         val itemset = items.toSet
-        brRules.value.flatMap(rule =>
-          if (items != null && rule._1.forall(item => itemset.contains(item))) {
-            rule._2.filter(item => !itemset.contains(item))
-          } else {
-            Seq.empty
-          }).distinct
+        brRules.value.filter(_._1.forall(itemset.contains))
+          .flatMap(_._2.filter(!itemset.contains(_))).distinct
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: are we aware of code changes here

Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: so we don't need to handle null item?

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: Hi @felixcheung , can you be more specific about the null item that concerns you? Thanks.

Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: right, 2 things - first just calling out while the PR says doc changes there is this one code change here.
second, before this code was checking `items != null` do we need not consider that now?

File: docs/ml-frequent-pattern-mining.md (hunk 15 above)
Comment (hhbyyh) [inline docs/ml-frequent-pattern-mining.md]: Please refer to https://issues.apache.org/jira/browse/SPARK-19899

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 17 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: `items != null` is already checked at two lines above. 
Please refer to the comments in the PR for the history of the code change. I can update title to include the code change.

Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: let's update the PR/JIRA if code change is required for the doc change.
otherwise, let's leave code change as a separate PR?

File: docs/ml-frequent-pattern-mining.md (hunk 15 above)
Comment (felixcheung) [inline docs/ml-frequent-pattern-mining.md]: I mean style it as code with backtick

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 16 above)
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: ping

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 17 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: I guess that's the right way. I will revert the code change.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala (hunk 16 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala]: got it.

File: docs/ml-frequent-pattern-mining.md (hunk 15 above)
Comment (hhbyyh) [inline docs/ml-frequent-pattern-mining.md]: sure.

**Comment by hhbyyh** [comment]
@felixcheung, reverted the code change of `transform` as requested. Please check the update. Thanks.

**Comment by jkbradley** [comment]
LGTM, thanks for adding this!  @felixcheung OK with merging?

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
merged to master/2.2
